cpp_common/src/lib.rs

```diff
@@ -89,23 +101,67 @@ impl ClosureSig {
     }
 
     pub fn extern_name(&self) -> Ident {
-        format!("__cpp_closure_{}", self.name_hash()).into()
+        Ident::new(
+            &format!("__cpp_closure_{}", self.name_hash()),
+            Span::call_site(),
+        )
     }
 }
 
-#[derive(Clone, Debug, PartialEq, Eq)]
+#[derive(Clone, Debug)]
 pub struct Closure {
     pub sig: ClosureSig,
-    pub body: Spanned<String>,
+    pub body: TokenTree,
+    pub body_str: String, // with rust! macro replaced
     pub callback_offset: u32,
 }
 
+impl Parse for Closure {
+    fn parse(input: ParseStream) -> Result<Self> {
+        input.parse::<Option<Token![unsafe]>>()?;
+        let cap_con;
+        bracketed!(cap_con in input);
+        let captures = syn::punctuated::Punctuated::<Capture, Token![,]>::parse_terminated(
+            &cap_con,
+        )?.into_iter()
+            .collect();
+        let (ret, cpp) = if input.peek(Token![->]) {
+            input.parse::<Token![->]>()?;
+            let t: syn::Type = input.parse()?;
+            input.parse::<Token![as]>()?;
+            let s = input.parse::<syn::LitStr>()?.value();
+            (Some(t), s)
+        } else {
+            (None, "void".to_owned())
+        };
+        let body = input.parse::<TokenTree>()?;
+        // Need to filter the spaces because there is a difference between
+        // proc_macro2 and proc_macro and the hashes would not match
+        let std_body = body
+            .to_string()
+            .chars()
+            .filter(|x| *x != ' ' && *x != '\n')
```

Comment:
We could probably use 'is_whitespace' here. It might also be nice to do this filtering inside of `ClosureSig` instead, and have a single entry point.